### PR TITLE
Implement Bandwidth Extension

### PR DESCRIPTION
**Pull request by WeberJulian**

### CODE REVIEW

File: TTS/enhancer/layers/losses.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import torch
+from TTS.utils.audio import TorchSTFT
+from TTS.tts.layers.losses import L1LossMasked
+from TTS.vocoder.layers.losses import MelganFeatureLoss, MSEGLoss, MSEDLoss, _apply_G_adv_loss, _apply_D_loss
+
+class BWEGeneratorLoss(torch.nn.Module):
+    def __init__(self, device):
+        super().__init__()
+        self.l1_masked= L1LossMasked(False)
+        self.specs = [TorchSTFT(
```
Comment (Edresson) [inline TTS/enhancer/layers/losses.py:11]: Maybe you can replace it with the loss_mag of the MultiScaleSTFTLoss, in this way you can reuse the code  and have the loss normalized.

File: TTS/enhancer/layers/losses.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        y_specs = [self.specs[i](y[:, 0, :]) for i in range(4)]
+        y_mel = self.mel_spec(y[:, 0, :])
+
+        y_hat_specs = [self.specs[i](y_hat[:, 0, :]) for i in range(4)]
+        y_hat_mel = self.mel_spec(y_hat[:, 0, :])
+
+        # Spectrogram loss
+        mel_lens = self.compute_lens(y_mel, lens)
+        return_dict["l1_mel"] = self.l1_masked(y_hat_mel, y_mel, mel_lens)
+        return_dict["loss"] = return_dict["l1_mel"]
```
Comment (Edresson) [inline TTS/enhancer/layers/losses.py:47]: You need to add with += not with = otherwise the l1_wavform loss will be ignored.

File: TTS/enhancer/datasets/dataset.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            audio_path = item["audio_file"]
+
+            # load wav file
+            target_wav = self.load_audio(audio_path)
+            # segment wav file
+            target_wav = self.segment_wav(target_wav)
+            # make sure that the length of the wav is a multiple of 3
+            if len(target_wav) %3 != 0:
+                target_wav = target_wav[:-(len(target_wav) % 3)]
+            input_wav = resample(target_wav, orig_sr=self.target_sr, target_sr=self.input_sr, res_type="zero_order_hold")
```
[the remark this answers is not in the document]
Comment (WeberJulian) [reply TTS/enhancer/datasets/dataset.py:107]: Yeah same results and since I don't need the gradients it was more convenient to use librosa

File: TTS/enhancer/layers/losses.py (hunk 1 above)
Comment (WeberJulian) [reply TTS/enhancer/layers/losses.py:11]: Good idea, I'll try that

File: TTS/enhancer/layers/losses.py (hunk 2 above)
Comment (WeberJulian) [reply TTS/enhancer/layers/losses.py:47]: Good catch!

File: TTS/enhancer/models/bwe.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+        if optimizer_idx not in [0, 1]:
+            raise ValueError(" [!] Unexpected `optimizer_idx`.")
+
+        if optimizer_idx == 0:
+            y_hat = self.gen_forward(x)["y_hat"]
+            self.y_hat_g = y_hat
+            scores_fake, feats_fake, feats_real = None, None, None
+            if self.train_disc:
+                scores_fake, feats_fake = self.disc_forward(y_hat)
```
Comment (Edresson) [inline TTS/enhancer/models/bwe.py:159]: I think that you need to detach the y_hat here.

File: TTS/enhancer/models/bwe.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+                with torch.no_grad():
+                    _, feats_real = self.disc_forward(y)
+            loss_dict = criterion[optimizer_idx](y_hat, y, lens, scores_fake, feats_fake, feats_real)
+            outputs = {"y_hat": y_hat}
+
+        if optimizer_idx == 1:
+            if self.train_disc:
+                y_d = y.clone()
+                y_hat = self.y_hat_g
+                scores_fake, feats_fake = self.disc_forward(y_hat.detach())
```
Comment (Edresson) [inline TTS/enhancer/models/bwe.py:169]: here is not necessary detach

File: TTS/enhancer/models/bwe.py (hunk 4 above)
Comment (Edresson) [reply TTS/enhancer/models/bwe.py:159]: my bad, the  optimizer_idx ==  0 is the generator optimizer, so it is right. I thought that it was the discriminator optimizer.

File: TTS/enhancer/models/bwe.py (hunk 5 above)
Comment (Edresson) [reply TTS/enhancer/models/bwe.py:169]: same here, my bad the code is right.

File: TTS/enhancer/models/bwe.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+        state = torch.load(checkpoint_path, map_location=torch.device("cpu"))
+        self.load_state_dict(state["model"], strict=strict)
+        if eval:
+            self.eval()
+            assert not self.training
+
+    def get_optimizer(self) -> List:
+        gen_params = list(self.generator.parameters()) + list(self.postconv.parameters())
+        optimizer_gen = get_optimizer(self.config.optimizer, self.config.optimizer_params, self.config.lr, parameters=gen_params)
+        disc_params = list(self.waveform_disc.parameters())
```
Comment (WeberJulian) [inline TTS/enhancer/models/bwe.py:241]: I think I found the issue...
